doc/source/conf.py

```diff
@@ -46,29 +44,79 @@
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
 # ones.
 extensions = [
-    # builtin extansions
-    "sphinx.ext.autodoc",
-    "sphinx.ext.autosummary",
-    "sphinx.ext.todo",
-    "sphinx.ext.mathjax",
-    "sphinx.ext.viewcode",
-    "numpydoc",
-    "nbsphinx",
-    "sphinx_autodoc_typehints",
+    "sphinx.ext.autodoc",  # Core Sphinx library for auto html doc generation from docstrings
+    "sphinx.ext.autosummary",  # Create neat summary tables for modules/classes/methods etc
+    "sphinx_autopackagesummary",  # Automatically search packages instead of hard-naming modules
+    "sphinx.ext.intersphinx",  # Link to other project's documentation (see mapping below)
+    "sphinx.ext.mathjax",  # Render math via Javascript
+    "sphinx.ext.viewcode",  # Add a link to the Python source code for classes, functions etc.
+    "sphinx_autodoc_typehints",  # Automatically document param types (less noise in class signature)
+    "sphinx.ext.todo",  # Support for to-do items
+    "nbsphinx",  # Integrate Jupyter Notebooks and Sphinx
     "IPython.sphinxext.ipython_console_highlighting",
 ]
 
-set_type_checking_flag = True
-typehints_fully_qualified = False
-always_document_param_types = True
-# autoclass_content = 'both'
+# Mappings for sphinx.ext.intersphinx. Projects have to have Sphinx-generated doc (.inv file)
+intersphinx_mapping = {
```

Comment:
Intersphinx seems quite useful to me - it means that when you reference a third party object from Python or Numpy then the documentation automatically links through to the Python/Numpy documentation for that object. It's a nice to have, I guess, rather than essential - but it makes for a nice user experience.

Unfortunately, the .inv file for Tensorflow, the library that gets referenced a lot, is non-functional at the moment, and the original developer is unresponsive. It may be possible to build and host the .inv file for TF ourselves, or we could just remove the TF mapping.

Reply:
> Personally, I'm not convinced that the individual multiple dispatch functions are really considered public API, since many have leading underscore and the whole point of multiple dispatch is to have one API to get one thing done, even if the implementations differ.

I would agree with this. My original plan was to write some custom words in the `__init__.py` file for each of the modules that uses multiple dispatch summarising what they were actually doing and include it in this PR, but I ran out of time.

It seems a shame to write custom summaries to document an API when a combination of Sphinx and the docstrings should be doing it automatically for us, but I would argue that even now, the current documentation doesn't do a great job of explaining what's going on in the dispatch modules. There's a lot of noise and it's hard to consume.